src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParserStateObject.cs

```diff
@@ -2383,19 +2448,33 @@ internal void ReadSni(TaskCompletionSource<object> completion)
             {
                 Debug.Assert(completion != null, "Async on but null asyncResult passed");
 
-                if (_networkPacketTimeout == null)
+                // if the state is currently stopped then change it to running and allocate a new identity value from 
+                // the identity source. The identity value is used to correlate timer callback events to the currently
+                // running timeout and prevents a late timer callback affecting a result it does not relate to
+                int previousTimeoutState = Interlocked.CompareExchange(ref _timeoutState, TimeoutRunning, TimeoutStopped);
+                if (previousTimeoutState == TimeoutStopped)
                 {
-                    _networkPacketTimeout = ADP.UnsafeCreateTimer(
-                        new TimerCallback(OnTimeout),
-                        null,
-                        Timeout.Infinite,
-                        Timeout.Infinite);
+                    Debug.Assert(_timeoutIdentityValue == 0, "timer was previously stopped without resetting the _identityValue");
+                    _timeoutIdentityValue = Interlocked.Increment(ref _timeoutIdentitySource);
                 }
 
+                _networkPacketTimeout?.Dispose();
+                TimeoutState state = Interlocked.Exchange(ref _spareTimeoutState, null) ?? new TimeoutState();
+                state.IdentityValue = _timeoutIdentityValue;
+                _networkPacketTimeout = ADP.UnsafeCreateTimer(
```

Comment:
The timeout exists only for a single read packet. I read through the runtime implementation to see if there was simple way for us to do slightly naughty things like private reflection but decided that there was no easy or even mildly tricky way to get what we need. Having debugged through it a couple of times it's clear that even when the timer is disposed if the continuation is active but unscheduled it can still come back.  As I said in the other thread I'm not wild about creating the timer every iteration but we need to do it and we've made enough speed and memory improvements elsewhere that the hit isn't going to cause anyone a problem, in my opinion.

The flow through ReadSni is tricky but it's mostly an initiator and you end up dealing with the actual read in MarsConnection.HandleReceiveComplete (or non-mars equivalent) It only cleans up in cases where there is a sync error known. 

Each receive is independent at this layer because it's just dealing with packets and the timeout calculations were already present and are unchanged so multipacket results will function as they did previously.

Reply:
@David-Engel 
I did try the same SqlCommand re-use and since we have an assigned "Owner", it's target is refreshed with re-execution and we are able to compare `_owner.target` appropriately.

For this PR, I'm not sure if it makes any additional improvement than what we got in #907 and looks more complex to understand. I'm trying out more cases and we're reviewing both PRs carefully.

Reply:
The owner should change but since it depends on the caller if that caller is the same at some point we'll accidentally get the same one twice in a row. At which point we end up with this bug again but even more difficult to catch and track down. Lets squash it completely now.

The difference between them is the additional synchronous timeout checking which makes timeouts more reliable in thread exhaustion conditions.

Reply:
So the reason why same owner is not a problem is because we rely on owner.target here, which always updates if you execute a new command, and owner changes if you execute same command twice. I'm interested to see in which can reproduce our thoughts with owner logic, as things are completely open now and we have all we need to reproduce anything. Let me know if you can make a situation that makes owner logic fail, and is unreliable.

Otherwise, I'm also very interested in timeout state and making it run in sync. Is it possible to simplify this PR by implementing these:
- Remove identity related counter that increments and use owner.target instead (if we can confirm it's reliability).
- Create and use Timeout state as enums.
- Simplify as much as possible.

I will also push my tests to this PR (will create a PR over it) that will likely cover many cases (work in progress), let's come up with a final PR with best solution we can get here!

Reply:
> Create and use Timeout state as enums.

This can't be done because enums are a value type but not seen by the compiler as their base type so there is no overload of CompareExchange which can work with them. CompareExchange is needed to prevent multiple paths executing. I started to do this and then found the problem so went with integers instead. I asked about it in the runtime discord and the reason is that it wasn't possible on framework but might be on core with the new enum generic constraint, however it doesn't exist now and there is no portable way to do it, the only real suggestion was to use `Unsafe.As<>` to do il level reinterpretation of the ref type and I don't think that's appropriate in this case.

> Simplify as much as possible.

This is the simple version. What is it that you think isn't needed?

> Remove identity related counter that increments and use owner.target instead (if we can confirm it's reliability).

If you use the owner you're going to have to use weakreference to stop a timer rooting a disposed object in weird situations. Which is fine if that's what is needed. I don't see a problem with using the counter though. It's probably unique over the full range of a 32 bit value and there are enough integers between the first and last that i'm not too worried about the first and last colliding, if that'a a concern we can go to a 64 bit integer but i didn't think that was needed given how rarely the problem actually manifestss. What is it about the counter that you dislike?

Reply:
> What is it about the counter that you dislike?

The state changes took me some time to digest. I'm pretty sure that's what Cheena is referring to. I'm okay with them at this point, but on first look, they were a bit hard to follow. I actually prefer the counter method at this point because I don't have to wonder if the identity for a timer has changed each time it has been created.

Reply:
@Wraith2 

> This can't be done because enums are a value type but not seen by the compiler as their base type so there is no overload of CompareExchange which can work with them.

A typical pattern is to wrap an int backing field with an enum property, which just converts back and forth (int to enum conversions are elided at runtime); the interlocked operation can be done directly on the backing field.

Reply:
If that isn't confusing then it might be worth a try. Can you point me at an example?

Reply:
[This is an example](https://github.com/npgsql/npgsql/blob/main/src/Npgsql/NpgsqlConnector.cs#L398) though looking at this a few years later, the concurrency control isn't good. If this is concurrency-sensitive, I think the backing field should be made volatile, and the interlocked operation can be removed.

Reply:
For the timeout state: Volatile would mean observers got the current value but then there is a time-of-check-time-of-use problem in that another thread can have modified the state that was just checked before the first thread modifies it, then you've got two threads that think they're the one that changed the state and we could end up with two opposing operations happening at once.
Interlocked ensures all reads and writes are synced over all threads/processors. There must be exactly one winner in every state of change and that winner must correctly know the starting state otherwise the transition is not safe. Interlocked does that. 

For the id variable:
I chose volatile for the id variable because it's only assigned once and that assigner knows the new value anyway. We only need to ensure that the other observer doesn't end up with a stale copy which volatile is good for.

In both cases I decided against locks because I don't want to wait on either side. Either it is the winner of the race to update the state or it isn't and in both cases it needs to quickly move on to avoid stalling.

Using an Enum which I would have preferred changes the common code from something like:
`Interlocked.Exchange(ref _timeoutState, TimeoutStopped);`
to 
`Interlocked.Exchange(ref _timeoutState, (int)TimeoutState.Stopped);`
which I don't think adds much in terms of readability and adds nothing in terms of compiled functionality. I can change it if that's what people prefer but I didn't see it as useful.
The interlocks are absolutely needed so CompareExchange isn't negotiable and volatile isn't strong enough so if it needs to be an enum then we'd have to drop down to lock(), again possible but I decided against it for the reasons outlined above.

Given the constant use throughout the library of weird stuff like `TdsEnums` ( which is a static class containing all sorts of variables of many types and non-interchangable uses that can only be deduced from context) I can appreciate the need for clear code and I've done the best I can given the technical constraints. I didn't just chose to write slightly strange non-idiomatic code there are good reasons for the choices (apart from the thread sleep, that was an error form debugging.

Reply:
Thanks @Wraith2 I don't see any obvious issues, and we can definitely go with your recommendations + what everyone approves on. I'll leave you guys to decide best design here :)

From testing perspective, I see this PR being more stable, as it handles Timeout Synchronously in case of delays and that makes a more consistent design. I'll let @Wraith2 merge https://github.com/Wraith2/SqlClient/pull/1 on to this one to add all tests, they seem to pass locally.

Reply:
> Volatile would mean observers got the current value but then there is a time-of-check-time-of-use problem in that another thread can have modified the state that was just checked before the first thread modifies it, then you've got two threads that think they're the one that changed the state and we could end up with two opposing operations happening at once.
> Interlocked ensures all reads and writes are synced over all threads/processors. There must be exactly one winner in every state of change and that winner must correctly know the starting state otherwise the transition is not safe. Interlocked does that.

If you need to do an atomic compare-and-swap, then you indeed need to use interlocked - I wasn't implying that volatile solves that problem. You would perform that operation directly against the int backing field, as you wrote above. However, other accesses which don't require atomicity - e.g. just verifying that you're at a certain state - can safely go through the property, and therefore use regular int values.

Reply:
Re Interlocked.Exchange - which you propose above - that really is useful only you need to get the previous value out atomically, in the same operation as you're putting the new value in. If you're not consulting the return value of Interlocked.Exchange, then there shouldn't be a reason to use that (since you're not caring about the 2nd part of the atomic Exchange operation). If the point is simply to assign the timeout state without any checks or anything, you can simply assign directly (with volatile).

Reply:
The exchange on spare was just to make sure it atomically took an object so we couldn't end up with two threads trying to re-use a single object. I realised that my logic around spares was incorrect and removed it. 

The important interlocks are all CompareExchange and those all have their return values checked before proceeding. The read use in `IsTimeoutStateExpired` isn't interlocked because it's just checking. There is one Exchange in `SetTimeoutStateStopped` which as you say really only serves as a read fence but since you can't CompareExchange on a volatile it felt right to make all writes atomic because it's a lot simpler to show anyone who reads the code a consistent pattern of assignment than it is to make them read this conversation and explain the subtleties involved.

So I think we agree on everything here apart from that last exchange and that's not incorrect just safer than might be needed.

Reply:
Yep, it all sounds good to me.